src/coreComponents/linearAlgebra/interfaces/petsc/PetscSuperLU_Dist.cpp

```diff
@@ -0,0 +1,165 @@
+{
+
+void PetscConvertToSuperMatrix( PetscMatrix const & matrix,
+                                Mat & localMatrix,
+                                SuperLU_Dist & SLUDData )
+{
+  GEOSX_LAI_CHECK_ERROR( MatMPIAIJGetLocalMat( matrix.unwrapped(), MAT_INITIAL_MATRIX, &localMatrix ) );
+
+  PetscInt numRows;
+  const PetscInt * ia;
+  const PetscInt * ja;
+  PetscBool status;
+  GEOSX_LAI_CHECK_ERROR( MatGetRowIJ( localMatrix, 0, PETSC_FALSE, PETSC_FALSE, &numRows, &ia, &ja, &status ) );
+  if( !status )
+  {
+    GEOSX_ERROR( "ConvertPetscToSuperMatrix: MatGetRowIJ reported an error." );
+  }
+
+  real64 * array;
+  GEOSX_LAI_CHECK_ERROR( MatSeqAIJGetArray( localMatrix, &array ) );
+
+  globalIndex const numGlobalRows = matrix.numGlobalRows();
+  localIndex const numLocalRows = matrix.numLocalRows();
+  localIndex const numLocalNonzeros = matrix.numLocalNonzeros();
+
+  SLUDData.setNumGlobalRows( LvArray::integerConversion< int_t >( numGlobalRows ) );
+  SLUDData.setNumLocalRows( LvArray::integerConversion< int_t >( numLocalRows ) );
+
+  SLUDData.createRowPtr( numLocalRows );
+  for( localIndex i = 0; i <= numLocalRows; ++i )
+  {
+    SLUDData.rowPtr()[i] = LvArray::integerConversion< int_t >( ia[i] );
+  }
+  SLUDData.createColIndices( numLocalNonzeros );
+  SLUDData.createValues( numLocalNonzeros );
+  for( localIndex i = 0; i < numLocalNonzeros; ++i )
+  {
+    SLUDData.colIndices()[i] = LvArray::integerConversion< int_t >( ja[i] );
+    SLUDData.values()[i] = array[i];
+  }
+  GEOSX_LAI_CHECK_ERROR( MatRestoreRowIJ( localMatrix, 0, PETSC_FALSE, PETSC_FALSE, &numRows, &ia, &ja, &status ) );
+  if( !status )
+  {
+    GEOSX_ERROR( "ConvertPetscToSuperMatrix: MatGetRowIJ reported an error." );
+  }
+  GEOSX_LAI_CHECK_ERROR( MatSeqAIJRestoreArray( localMatrix, &array ) );
+
+  dCreate_CompRowLoc_Matrix_dist( &SLUDData.mat(),
+                                  toSuperLU_intT( numGlobalRows ),
+                                  toSuperLU_intT( numGlobalRows ),
+                                  toSuperLU_intT( numLocalNonzeros ),
+                                  toSuperLU_intT( numLocalRows ),
+                                  toSuperLU_intT( matrix.ilower() ),
+                                  SLUDData.values(),
+                                  SLUDData.colIndices(),
+                                  SLUDData.rowPtr(),
+                                  SLU_NR_loc,
+                                  SLU_D,
+                                  SLU_GE );
+
+  SLUDData.setComm( matrix.getComm() );
+}
+
+void PetscDestroyAdditionalData( Mat & localMatrix )
+{
+  GEOSX_LAI_CHECK_ERROR( MatDestroy( &localMatrix ) );
+}
+
+namespace
+{
+class InverseOperator
+{
+public:
+
+  ~InverseOperator()
+  {
+    PetscDestroyAdditionalData( m_localMatrix );
+  }
+
+  void set( PetscMatrix const & matrix, SuperLU_Dist & SLUDData )
+  {
+    m_SLUDData = &SLUDData;
+    m_comm = SLUDData.getComm();
+
+    matrix.transpose( m_transposeMatrix );
```

Comment:
Doesn't SuperLU_Dist have a transpose solve option? Could you use it to get rid of explicit matrix transpose?

Reply:
You're right! Yes, SuperLU_Dist provides a transpose solver, but it does not work if coupled with row/col scaling!
Disabling the two scalings, it works, but for a general matrix the scalings are needed, thus I preferred to do all the computations on our side.

Reply:
Fair enough, totally good reason to transpose on our side